### DISCUSSION TITLE
Separate concurrency limits with CeleryKubernetes executor

### DISCUSSION THREAD

**Discussion by rcheatham-q** [apache/airflow#31305] [Q&A]
When running the CeleryKubernetes executor I would like to limit the number of tasks executing in each executor separately. Right now it seems I have to set the limits across both executors. Is there a way to do this with the current configuration options? Using pools is not acceptable because I have more specific pools for the various services my instance interacts with. For example, I can only have 20 Spark applications running concurrently so I have a `spark` pool of size 20, but there are other tasks I would like to run in the Celery workers and I don't want to decrease the overall pool for Celery workers by 20 because I rarely have 20 Spark jobs running.

**Comment by kbumsik** [comment]
AFAIK Kubernetes Executor doesn't support configurable concurrency though. You run 1 pod per job.

**Reply by rcheatham-q** [reply]
You can configure it using the core.parallelism config option, which is global. Celery is only marginally better with the celery.worker_concurrency option, however that doesn't prevent the scheduler from queueing tasks which is what I'm hoping for.

**Reply by potiuk** [reply]
Yes. Concurrency is based on different rules in K8S part. K8S executor has no limits - it will schedule as many pods as many there are eligible (taking into pools and queues, and other limits that you can set per dag and tasks) and it's up to K8S to scheduler to run them depending on how many resources are available - so you can effectively manage concurency by specifying resource requests in each of the POd that gets scheduled - for example if each of the PODS of yours has 100 MB memory requirement and the whole cluster has free 10GB memory then (give or take depending on how busy your cluster is) you will be abe to have 100 tasks running in total for K8S executor. You can do the same with other resources - making it much more flexible (for example memory hungry tasks can have bigger requests)

**Reply by potiuk** [reply]
You can use for example:

* `max_active_tasks` in DAG  – the number of task instances allowed to run concurrently
 
for DAG or 

* `max_active_tis_per_dag` ([int](https://docs.python.org/3/library/functions.html#int) | None) – When set, a task will be able to limit the concurrent runs across execution_dates.
* `max_active_tis_per_dagrun` – When set, a task will be able to limit the concurrent task instances per DAG run.

for particular tasks

What problem are you trying to solve by not changing the tasks to "queued" I wonder?

**Comment by rcheatham-q** [comment]
> What problem are you trying to solve by not changing the tasks to "queued" I wonder?

I have found during extensive testing that the Kubernetes executor does not play well when tasks remain in the queued state for very long. The behavior manifests as random tasks failing which Airflow reports as an external task kill. Digging through the logs for the scheduler I believe there is a race condition: the executor creates the pod, then the scheduler decides to reschedule the task, then the pod starts, then the scheduler finds the task in a state it doesn't expect and kills it. I found the executor functions far better if I can prevent the tasks from entering the queued state until I can be sure that there is enough room in the cluster to create the pod.

**Reply by rcheatham-q** [reply]
I should clarify something: I am not admin of my Kubernetes cluster. I have access to a single namespace that has a resource quota applied. When the Kubernetes API rejects a pod creation request it reports a 403 forbidden status code. Digging into the code it seems like this should be handled appropriately, however it is definitely not. I have tested using Celery and there were no issues with leaving tasks in the queued state for an extended period of time.

**Reply by potiuk** [reply]
Then you should raise an issue with your observations logs, version of Airlfow, K8s etc. - this might be bug, or it might have been saved in newer Airflow version (for one there was a problem with queued tasks solved in Airflow 2.6.0 )

**Reply by rcheatham-q** [reply]
Okay, I'll see if the issue is present in 2.6.1 and create an issue if it still exists.

**Reply by rcheatham-q** [reply]
Looks like this problem still exists and if anything is worse than it was before. I'll work on gathering logs etc. and submitting a bug report.
